### PR TITLE
rust: add event file reading module

### PR DESCRIPTION
**Pull request by wchargin**
Summary:
We implement a module for reading a TensorFlow event file. This wraps
the TFRecord reading module added in #4307 and parses records as protos.
The reader also stores the last-read event wall time, so that we can
stop reading from event files after a while (cf. `--reload_multifile`).

Along the way, we add a couple utility methods to `TfRecord` that make
it easier to use records in test code.

Test Plan:
Unit tests included.

wchargin-branch: rust-event-file-reading

### CODE REVIEW

**Comment by wchargin** [comment]
Hi! Here’s a code review to get us started. Diffstat says +300, but that
breaks down to:

  - `event_file.rs`: 65 lines of code, 95 lines of test, 34 lines of docs
  - `tf_record.rs`: 12 lines of code, 38 lines of test, 9 lines of docs

…so I hope it’s not too intimidating. As always, happy to discuss or
take any questions.

File: tensorboard/data/server/event_file.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        assert_eq!(reader.last_wall_time(), &None);
+        assert_eq!(reader.read_event().unwrap(), good_event);
+        assert_eq!(reader.last_wall_time(), &Some(1234.5));
+        match reader.read_event() {
+            Err(ReadEventError::NanWallTime(mut e)) => {
+                // can't just check `e == nan_event` because `NaN != NaN`
+                assert!(e.wall_time.is_nan());
+                e.wall_time = 0.0;
+                nan_event.wall_time = 0.0;
+                assert_eq!(e, nan_event);
```
Comment (psybuzz) [inline tensorboard/data/server/event_file.rs:167]: It makes sense after actually reading the comment on L163 :)

File: tensorboard/data/server/tf_record.rs [hunk 2]
```diff
@@ -352,4 +383,46 @@ mod tests {
...
+        }
+    }
+
+    fn test_write_read_roundtrip(record: &TfRecord) {
+        let mut cursor = Cursor::new(Vec::<u8>::new());
+        record.write(&mut cursor).expect("failed to write record");
+        let written_len = cursor.position();
+        cursor.set_position(0);
+        let mut reader = TfRecordReader::new(cursor);
+        let output_record = reader.read_record().expect("read_record");
```
Comment (psybuzz) [inline tensorboard/data/server/tf_record.rs:405]: nit: `...expect("read_record");` --> `expect("failed to read record");` ?

File: tensorboard/data/server/event_file.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+use std::io::Read;
+
+use crate::proto::tensorboard::Event;
+use crate::tf_record::{ChecksumError, ReadRecordError, TfRecordReader};
+
+/// A reader for a stream of `Event` protos framed as TFRecords.
+///
+/// As with [`TfRecordReader`], an event may be read over one or more underlying reads, to support
+/// growing, partially flushed files.
+pub struct EventFileReader<R> {
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/server/event_file.rs:29]: Done; thanks.

File: tensorboard/data/server/tf_record.rs [hunk 4]
```diff
@@ -128,6 +154,11 @@ impl<R: Read> TfRecordReader<R> {
         }
     }
 
+    /// Consumes this `TfRecordReader<R>`, returning the underlying reader `R`.
+    pub fn into_inner(self) -> R {
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/server/tf_record.rs:179]: Yeah. You can imagine something like:

```rust
// parse a file that has one TFRecord as a header and then just a bunch of data
let mut f = File::open("...")?;
let mut reader = EventFileReader::new(f);
let header = reader.read_event()?;
let mut f = reader.into_inner(); // give me the normal file back
let mut buf = Vec::new();
f.read_to_end(&mut buf)?;
```

File: tensorboard/data/server/tf_record.rs (hunk 2 above)
Comment (wchargin) [reply tensorboard/data/server/tf_record.rs:426]: I’m not certain what the prevailing style is here, but it renders like
this:

```
thread '<test_name>' panicked at 'read_record: BadLengthCrc(Truncated)', tensorboard/data/server/tf_record.rs:123:45
```

so I think it’s already clear enough. The fact that it’s panicking and
the test fails means that it’s failed.

If this were in non-test code and I were using `expect` for a condition
that _should_ never happen (i.e., an assertion), I’d be more careful to
provide clear information and context. But for a test that’s completely
self-contained, I think that this is fine.

File: tensorboard/data/server/event_file.rs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    /// Reads the next event from the file.
+    pub fn read_event(&mut self) -> Result<Event, ReadEventError> {
+        let record = self.reader.read_record()?;
+        let event = match Event::decode(&record.data[..]) {
+            Ok(ev) => ev,
+            Err(err) => {
+                // On proto decoding failure, check the record checksum first.
+                record.checksum()?;
```
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/data/server/event_file.rs]: Sure; that’s reasonable. Done, with test.

**Review by psybuzz — APPROVED** [review]
lgtm, thanks for walking me through this!

**Pull request merged**